projects/bootloader/src/query.c

```diff
@@ -0,0 +1,151 @@
+#include "query.h"
+
+#include <pb_common.h>
+#include <pb_decode.h>
+#include <pb_encode.h>
+#include <string.h>
+
+#include "bootloader_can.h"
+#include "bootloader_datagram_defs.h"
+#include "can_datagram.h"
+#include "config.h"
+#include "dispatcher.h"
+#include "querying.pb.h"
+#include "querying_response.pb.h"
+
+#define MAX_STRING_SIZE 64
+#define PROTOBUF_MAXSIZE 270  // enough to fit the max possible encoded protobuf
+#define NUM_QUERY_FIELDS 5
+
+typedef enum {
+  NOT_FOUND = 0,
+  FOUND,
+  NON_EXISTANT,
+} CompareResult;
+
+static void *s_targets[NUM_QUERY_FIELDS];
+static CompareResult s_results[NUM_QUERY_FIELDS];
+static Querying s_querying = Querying_init_zero;
+
+static uint8_t s_datagram_data[PROTOBUF_MAXSIZE];
+
+static CanDatagramTxConfig s_response_config = {
+  .dgram_type = BOOTLOADER_DATAGRAM_QUERY_RESPONSE,
+  .destination_nodes_len = 0,  // client listens to all datagrams
+  .destination_nodes = NULL,
+  .data_len = 0,
+  .data = s_datagram_data,
+  .tx_cb = bootloader_can_transmit,
+  .tx_cmpl_cb = tx_cmpl_cb,
+};
+
+// compare the int query field to the board field
+static bool prv_decode_cmp_varint(pb_istream_t *stream, const pb_field_iter_t *field, void **arg) {
+  if (s_results[field->index] == FOUND) {
+    return true;
+  }
+  s_results[field->index] = NOT_FOUND;
+
+  uint8_t *target = s_targets[field->index];
+  uint64_t incoming;
+  if (!pb_decode_varint(stream, &incoming)) {
+    return false;
+  }
+  if (incoming == *target) {
+    s_results[field->index] = FOUND;
+  }
+  return true;
+}
+
+// compares the query field to the board field directly without storying the query
+static bool prv_decode_cmp_string(pb_istream_t *stream, const pb_field_iter_t *field, void **arg) {
+  if (s_results[field->index] == FOUND) {
+    return true;
+  }
+  s_results[field->index] = NOT_FOUND;
+
+  char *target = (char *)s_targets[field->index];
+  if (target == NULL) {
+    return true;
+  }
+  // compare the field to the board field
+  if (strncmp(target, stream->state, stream->bytes_left) == 0) {
```

Comment:
To be extra safe you might want to pass `MIN(stream->bytes_left, MAX_STRING_SIZE)` as the third parameter to strncmp (`MIN` is from misc.h).